### PR TITLE
[HOPSWORKS-640] Flink 1.8.1

### PR DESCRIPTION
**Pull request by tkakantousis**

### CODE REVIEW

File: recipes/historyserver.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+end
+
+template node['flink']['historyserver']['environment'] do
+    source "historyserver.env.erb"
+    owner node['flink']['user']
+    group node['flink']['group']
+    mode 0750
+    variables({
+      :hadoop_glob=> lazy { node['hadoop_glob'] },
+      :flink_log_dir=> node['flink']['historyserver']['logs']
```
Comment (SirOibaf) [inline recipes/historyserver.rb]: You can avoid passing the the `flink_log_dir` as variable, just use `<%= node['flink']['historyserver']['logs'] %>`. Keeps the recipes more concise and readable

File: Karamelfile [hunk 2]
```diff
@@ -2,6 +2,7 @@ dependencies:
   - recipe: flink::install
     global:  
       - kagent::install
+      - hops::install
   - recipe: flink::taskmanager
```
Comment (SirOibaf) [inline Karamelfile]: Are the `flink::taskmanager` and `flink::jobmanager` recipes still valid?

File: Karamelfile [hunk 3]
```diff
@@ -2,6 +2,7 @@ dependencies:
   - recipe: flink::install
     global:  
       - kagent::install
+      - hops::install
```
Comment (SirOibaf) [inline Karamelfile]: Karamel 0.6 does not respect install dependencies.

File: recipes/historyserver.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    })
+    notifies :enable, resources(:service => service_name)
+    notifies :start, resources(:service => service_name), :immediately
+  end
+
+kagent_config service_name do
+  action :systemd_reload
+end
+
+if node['services']['enabled'] != "true" && node['flink']['systemd'] == "true"
```
Comment (SirOibaf) [inline recipes/historyserver.rb]: I don't think you need to check for `node['flink']['systemd']` the rest of the recipe and the entire platform relies on Systemd.

File: recipes/historyserver.rb [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    service "flinkhistoryservice" do
+      provider Chef::Provider::Service::Systemd
+      supports :restart => true, :stop => true, :start => true, :status => true
+      action :disable
+    end
+
+    kagent_config "flinkhistoryservice" do
```
Comment (SirOibaf) [inline recipes/historyserver.rb]: Copy paste left over

File: recipes/install.rb [hunk 6]
```diff
@@ -7,11 +7,33 @@
 
 include_recipe "java"
 
+
+group node['hops']['group'] do
```
Comment (SirOibaf) [inline recipes/install.rb]: Could you please explain why you need to create the `hdfs` user here?

File: templates/default/sdk_worker.sh.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+export HADOOP_HDFS_HOME=<%= node['hops']['base_dir'] %>
+export HADOOP_CLASSPATH=::/*:/lib/*
+export HADOOP_CONF_DIR=<%= node['hops']['conf_dir'] %>
+export CLASSPATH="<%= @hadoop_glob %>:$CLASSPATH"
+export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:${JAVA_HOME}/jre/lib/amd64/server
+export PATH=$PATH:${HADOOP_HOME}/bin
+
+SUBSTR=$(echo $USER | cut -d'_' -f 1)
+ 
+bash -c "source /srv/hops/anaconda/anaconda/bin/activate /srv/hops/anaconda/anaconda/envs/$SUBSTR && ./boot $* >> $LOG_DIRS/beamsdkworker-$SUBSTR-$CONTAINER_ID.log 2>&1 "
```
Comment (SirOibaf) [inline templates/default/sdk_worker.sh.erb]: Here you didn't consider that users might be running using the base environments

File: templates/default/flinkhistoryserver.service.erb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description = Flink historyserver
+After = syslog.target network.target remote-fs.target <%= @deps %>
+
+[Service]
+User = <%= node['hops']['hdfs']['user'] %>
```
Comment (SirOibaf) [inline templates/default/flinkhistoryserver.service.erb]: Why do we need to run the flinkhistoryserver as `hdfs`?

File: recipes/yarn.rb [hunk 9]
```diff
@@ -23,10 +23,40 @@
...
+
+ruby_block "hadoop_glob" do
+  block do
+    Chef::Resource::RubyBlock.send(:include, Chef::Mixin::ShellOut)
+    node.default['hadoop_glob'] = shell_out("#{node['hops']['bin_dir']}/hadoop classpath --glob  | tr -d '\n'").stdout
+  end
+  action :create
 end
+
+template "#{node['flink']['home']}/conf/sdk_worker.sh" do
```
Comment (SirOibaf) [inline recipes/yarn.rb]: Is this recipe supposed to be run on every machine (like `spark::yarn`)? if so, could you please wrap all the `hops_hdfs_directory` in an if?

File: Karamelfile (hunk 2 above)
Comment (tkakantousis) [reply Karamelfile]: No, but they template some options for flink-conf.yml I think, I will try to remove them in any case.

File: Karamelfile (hunk 3 above)
Comment (tkakantousis) [reply Karamelfile]: Sure, but it's nice to had it there anyway in case some day Karamel supports that

File: recipes/install.rb (hunk 6 above)
Comment (tkakantousis) [reply recipes/install.rb]: it happened that flink::install ran before hops::install and the user did not exist. I will move the code to using the `flink` user anyway.

File: templates/default/sdk_worker.sh.erb (hunk 7 above)
Comment (tkakantousis) [reply templates/default/sdk_worker.sh.erb]: You are correct, I wrote this code when `tfx` was not part of the base environment. I need to first check in the script if the project's environment exists. Also `SUBSTR=$(echo $USER | cut -d'_' -f 1)` will fail if the project name contains the `_` character, need to fix this as well.

File: recipes/yarn.rb (hunk 9 above)
Comment (tkakantousis) [reply recipes/yarn.rb]: No, flink is installed on one machine at the moment. Flink internally ships the `flink.jar` as a local resource. Now we added support in Flink for custom local resources that are stored in hdfs, so I use that for the `sdk_worker.sh` script.

File: templates/default/flinkhistoryserver.service.erb (hunk 8 above)
Comment (tkakantousis) [reply templates/default/flinkhistoryserver.service.erb]: We don't any more, will change it to `flink` user

File: recipes/historyserver.rb (hunk 4 above)
Comment (tkakantousis) [reply recipes/historyserver.rb]: Can we remove this check for other cookbooks then?

**Pull request merged**